src/library/dlgtrackinfo.cpp

```diff
@@ -468,46 +462,13 @@ void DlgTrackInfo::clear() {
     m_pWCoverArtLabel->setCoverArt(m_loadedCoverInfo, QPixmap());
 }
 
-void DlgTrackInfo::slotBpmDouble() {
-    m_pBeatsClone = m_pBeatsClone->scale(mixxx::Beats::DOUBLE);
-    // read back the actual value
-    double newValue = m_pBeatsClone->getBpm();
-    spinBpm->setValue(newValue);
-}
-
-void DlgTrackInfo::slotBpmHalve() {
-    m_pBeatsClone = m_pBeatsClone->scale(mixxx::Beats::HALVE);
-    // read back the actual value
-    double newValue = m_pBeatsClone->getBpm();
-    spinBpm->setValue(newValue);
-}
-
-void DlgTrackInfo::slotBpmTwoThirds() {
-    m_pBeatsClone = m_pBeatsClone->scale(mixxx::Beats::TWOTHIRDS);
-    // read back the actual value
-    double newValue = m_pBeatsClone->getBpm();
-    spinBpm->setValue(newValue);
-}
-
-void DlgTrackInfo::slotBpmThreeFourth() {
-    m_pBeatsClone = m_pBeatsClone->scale(mixxx::Beats::THREEFOURTHS);
-    // read back the actual value
-    double newValue = m_pBeatsClone->getBpm();
-    spinBpm->setValue(newValue);
-}
-
-void DlgTrackInfo::slotBpmFourThirds() {
-    m_pBeatsClone = m_pBeatsClone->scale(mixxx::Beats::FOURTHIRDS);
-    // read back the actual value
-    double newValue = m_pBeatsClone->getBpm();
-    spinBpm->setValue(newValue);
-}
-
-void DlgTrackInfo::slotBpmThreeHalves() {
-    m_pBeatsClone = m_pBeatsClone->scale(mixxx::Beats::THREEHALVES);
+void DlgTrackInfo::slotScaleBpm(const mixxx::Beats::BPMScale scale) {
+    if (!m_pBeatsClone) {
+        return;
+    }
+    m_pBeatsClone = m_pBeatsClone->scale(scale);
```

Comment:
this works around the issue, but it doesn't actually really fix it. Why is `m_pBeatsClone` nullptr when the bpm is 0?

Reply:
because it is cleared in various places, for example when the BPM spinbox value is changed to 0, or when the beatgrid is cleared by the Clear button. m_pBeatsClone.clear()

Reply:
This fix is valid.

Reply:
Ok thanks. I just wish this invariant was documented (that `m_pBeatsClone == nullptr` is a valid state of the class)